Adding code to create URLMaps

nikhiljindal opened the pull request:
This uses the backend services created in https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/pull/3 to create the required URL map.

cc @G-Harmon @csbell @madhusudancs @bowei @nicksardo

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/7)
<!-- Reviewable:end -->

nikhiljindal commented:
Verified that the tests are passing

```
$ go test ./...
ok      github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/cmd     0.354s
ok      github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/backendservice  0.229s
?       github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/cloudinterface  [no test files]
ok      github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/healthcheck     0.291s
ok      github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/loadbalancer    0.348s
?       github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/namer   [no test files]
ok      github.com/GoogleCloudPlatform/k8s-multicluster-ingress/app/mci/pkg/gcp/urlmap  0.200s
```

bowei commented:
@nikhiljindal it's very easy to add travis integration, take a look at https://github.com/kubernetes/dns/blob/master/.travis.yml

nikhiljindal commented:
Thanks @bowei Will take a look!

csbell commented on app/mci/pkg/gcp/backendservice/backendservicesyncer.go:
```diff
@@ -85,18 +89,43 @@ func (b *BackendServiceSyncer) ensureBackendService(lbName string, port ingressb
+	if err != nil {
+		return nil, err
+	}
+	fmt.Println("Backend service", name, "created successfully")
```
Unless you can guarantee that Get will never fail after a successful Create, it may be confusing to see "created successfully".

nikhiljindal replied:
Our cloud provider implementation waits for the operation to complete: https://github.com/kubernetes/kubernetes/blob/68d2722be072d229a347585fa7f964bf54b38c95/pkg/cloudprovider/providers/gce/gce_backendservice.go#L92.
So Get is guaranteed not to fail, unless of course you delete it first :)

csbell commented on app/mci/pkg/gcp/backendservice/backendservicesyncer_test.go:
```diff
@@ -39,10 +41,11 @@ func TestEnsureBackendService(t *testing.T) {
 		{
 			Port:     port,
 			Protocol: "HTTP",
+			SvcName:  types.NamespacedName{Name: kubeSvcName},
```
Non empty namespace pls?

nikhiljindal replied:
We always omit inconsequential fields in tests. Here namespace does not matter at all. None of the code tested in this unit test reads it.
I can add it if you really want, but that will create the perception that it matters.

csbell replied:
Got it. Not needed.

csbell commented on app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go:
```diff
@@ -98,10 +102,17 @@ func (l *LoadBalancerSyncer) CreateLoadBalancer(ing *v1beta1.Ingress, forceUpdat
+		err = multierror.Append(err, umErr)
+	}
+	return err
 }
```
How/where do we delete the healthcheck if, for example, we can successfully create the HC but the backendservice fails.

nikhiljindal replied:
Good point. We dont. User can retry with `mci create` again or delete resources by running `mci delete`.

csbell replied:
I'm not sure that this is a good user experience. Don't want to block the PR but let's file an issue to get back to this soon.

nikhiljindal replied:
Thx, filed https://github.com/GoogleCloudPlatform/k8s-multicluster-ingress/issues/8

G-Harmon commented:
Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[app/mci/pkg/gcp/backendservice/backendservicesyncer.go, line 47 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFC9mf005J-jrRPixO:-KxFByIurdbOoloawsUJ:bpnu8nf) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/backendservice/backendservicesyncer.go#L47)):*
> ```Go
> // EnsureBackendService ensures that the required backend services exist for the given ports.
> // Does nothing if it exists already, else creates a new one.
> func (b *BackendServiceSyncer) EnsureBackendService(lbName string, ports []ingressbe.ServicePort, hcMap healthcheck.HealthChecksMap, npMap NamedPortsMap, igLinks []string) (BackendServicesMap, error) {
> ```

update comment with new return type.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 64 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFDdTh_CBNSNCUa-4l:-KxFDdTh_CBNSNCUa-4m:b45wk25) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L64)):*
> ```Go
> )
> 
> type LoadBalancerSyncer struct {
> ```

add a comment for the overall struct?

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 108 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFDsI7tDv0Vdh3r8QV:-KxFDsI7tDv0Vdh3r8QW:bs42vc3) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L108)):*
> ```Go
> 	if beErr != nil {
> 		// Aggregate errors and return all at the end.
> 		err = multierror.Append(err, beErr)
> ```

question: Does it make sense to continue to EnsureUrlMap if there was an error in EnsureBackendService?

---

*[app/mci/pkg/gcp/urlmap/fake_urlmapsyncer.go, line 40 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFELIVzgN7Ncm-6SCH:-KxFELIVzgN7Ncm-6SCI:budoyc2) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/fake_urlmapsyncer.go#L40)):*
> ```Go
> 
> // Ensure this implements URLMapSyncerInterface.
> var _ URLMapSyncerInterface = &FakeURLMapSyncer{}
> ```

newbie question: is this like an ASSERT() to catch an error, or is it always required?

---

*[app/mci/pkg/gcp/urlmap/interfaces.go, line 25 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFEchgkAQCGexk4K_L:-KxFEchgkAQCGexk4K_M:b-7ni6e0) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/interfaces.go#L25)):*
> ```Go
> // URLMapSyncerInterface is an interface to manage GCP url maps.
> type URLMapSyncerInterface interface {
> 	// EnsureURLMap ensures that the required url map exists for the given ingress.
> ```

can you reference the BackendServicesMap in the comment? (at least, that'd be good for someone less familiar with the code)

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 34 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFJHjrLYXoUXM7pS1z:-KxFJHjrLYXoUXM7pS2-:b-qro7hs) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L34)):*
> // TODO(nikhiljindal): Refactor to share with kubernetes/ingress-gce
Seems like there's a lot of copied code from ingress-gce in this PR.  What's the plan/priority of refactoring to share that code?

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 35 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFEtL_YC8G1rXZc3WV:-KxFEtL_YC8G1rXZc3WW:b-bdnehf) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L35)):*
> ```Go
> 	// The gce api uses the name of a path rule to match a host rule.
> 	// TODO(nikhiljindal): Refactor to share with kubernetes/ingress-gce
> 	hostRulePrefix = "host"
> ```

Is this a dependency on ingress-gce? If so, can you refernce the exact file/function that it needs to be in sync with? (In google3, there's a cool presubmit check called IfThisThenThat or something, that allows you to require that both locations be updated in the same PR.)

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 38 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFFHsoJpTAxFvkVpSN:-KxFFHsoJpTAxFvkVpSO:b-kv5mhj) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L38)):*
> ```Go
> )
> 
> type URLMapSyncer struct {
> ```

short "class comment" on UrlMapSyncer would be good

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 79 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFFaZPdAVey9IlvoXp:-KxFFaZPdAVey9IlvoXq:b-2kpq0a) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L79)):*
> ```Go
> 		return h.updateURLMap(desiredUM)
> 	}
> 	glog.V(5).Infof("Got error %s while trying to get existing url map %s", err, name)
> ```

how about adding "... going ahead with creation anyway"

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 155 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFHtF3kNFkATpFoENe:-KxFHtF3kNFkATpFoENf:b-ldaoj1) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L155)):*
> subdomain: 
This is actually a set of nested maps, right? I think the comment, "map of subdomain : url-regex : gce backend", could be a little clearer. s/map/set of nested maps/.

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 169 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFHJuGCkILU5krW-5l:-KxFHJuGCkILU5krW-5m:b-4z7tqc) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L169)):*
> ```Go
> 			backend, err := getBackendService(&p.Backend, ing.Namespace, beMap)
> 			if err != nil {
> 				fmt.Println("Skipping path", p.Backend, "due to error", err)
> ```

Should we aggregate/return these errors? (if not, comment why not)

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 215 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFIjnEcPcqmNbd8QmO:-KxFIjnEcPcqmNbd8QmP:b8aki9e) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L215)):*
> TODO
add username

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[app/mci/pkg/gcp/backendservice/backendservicesyncer.go, line 47 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFC9mf005J-jrRPixO:-KxFN1mkDW8ZVQjxGDUg:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/backendservice/backendservicesyncer.go#L47)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

update comment with new return type.
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 64 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFDdTh_CBNSNCUa-4l:-KxFNdjSnVHhjyH2K9kF:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L64)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

add a comment for the overall struct?
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go, line 108 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFDsI7tDv0Vdh3r8QV:-KxFNqYi_wZ-7KF0zjoH:bgt3p9l) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/loadbalancer/loadbalancersyncer.go#L108)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

question: Does it make sense to continue to EnsureUrlMap if there was an error in EnsureBackendService?
</blockquote></details>

The aim is to return all errors at once, so if there are any errors in EnsureUrlMap find them out now.
EnsureBackendService can fail to create one backend service but succeed for others (incase of a single ingress pointing to multiple services). So we can try creating URL map for those services.

We can ofcourse change this if it doesnt make sense as per our usability testing and user feedback.

---

*[app/mci/pkg/gcp/urlmap/fake_urlmapsyncer.go, line 40 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFELIVzgN7Ncm-6SCH:-KxFOYmIA-mdMPTY8Ra5:b-8pr3op) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/fake_urlmapsyncer.go#L40)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

newbie question: is this like an ASSERT() to catch an error, or is it always required?
</blockquote></details>

This will generate a build time error if the fake syncer stops implementing the interface.

---

*[app/mci/pkg/gcp/urlmap/interfaces.go, line 25 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFEchgkAQCGexk4K_L:-KxFaxNJNHwnJioWpieP:b-896fix) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/interfaces.go#L25)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

can you reference the BackendServicesMap in the comment? (at least, that'd be good for someone less familiar with the code)
</blockquote></details>

Done.

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 34 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFJHjrLYXoUXM7pS1z:-KxFb4iVrZkoQqSBECVy:b-qmz5ow) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L34)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> // TODO(nikhiljindal): Refactor to share with kubernetes/ingress-gce
Seems like there's a lot of copied code from ingress-gce in this PR.  What's the plan/priority of refactoring to share that code?
</blockquote></details>

Yes exploring that in discussion with Bowei.
Some are not as complex and will do them right after we have the basic code working

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 35 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFEtL_YC8G1rXZc3WV:-KxFcRzWlSbTZimyK2kt:bh8u0x5) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L35)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Is this a dependency on ingress-gce? If so, can you refernce the exact file/function that it needs to be in sync with? (In google3, there's a cool presubmit check called IfThisThenThat or something, that allows you to require that both locations be updated in the same PR.)
</blockquote></details>

The file/function in ingress-gce can change.
Updated the comment to make it explicit that the constant exists with same name in ingress-gce so it is easy to search.

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 38 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFFHsoJpTAxFvkVpSN:-KxFfmqo0tzd5F1PTM9K:b4jchf4) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L38)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

short "class comment" on UrlMapSyncer would be good
</blockquote></details>

Sure, done

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 79 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFFaZPdAVey9IlvoXp:-KxFgDg66OGjr5aNVAGv:bb7f2mf) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L79)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

how about adding "... going ahead with creation anyway"
</blockquote></details>

done

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 155 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFHtF3kNFkATpFoENe:-KxFgcLVsWviliJZqVsd:buw47t2) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L155)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> subdomain: 
This is actually a set of nested maps, right? I think the comment, "map of subdomain : url-regex : gce backend", could be a little clearer. s/map/set of nested maps/.
</blockquote></details>

Updated the comment

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 169 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFHJuGCkILU5krW-5l:-KxFj_SdQuHdFcxElbxh:b-1a6peh) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L169)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

Should we aggregate/return these errors? (if not, comment why not)
</blockquote></details>

Good catch. Updated to return aggregated errors.

---

*[app/mci/pkg/gcp/urlmap/urlmapsyncer.go, line 215 at r1](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-KxFIjnEcPcqmNbd8QmO:-KxFjjIPKgKNiyOOQIyc:bb6un1j) ([raw file](https://github.com/googlecloudplatform/k8s-multicluster-ingress/blob/9274fdbb21dd281161a84b08cd27327d08acf237/app/mci/pkg/gcp/urlmap/urlmapsyncer.go#L215)):*
<details><summary><i>Previously, G-Harmon wrote…</i></summary><blockquote>

> TODO
add username
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7)*
<!-- Sent from Reviewable.io -->

G-Harmon commented:
LGTM
i don't see the new revision yet, but i didn't have any blocking comments.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/googlecloudplatform/k8s-multicluster-ingress/7#-:-KxFp6gp2ldPnOvSWVk2:b-5cbkg8)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Thanks for the super quick reply @G-Harmon!
Pushed the updated code as per comments.

csbell commented:
/lgtm
/approve
